File: app/src/main/java/org/docheinstein/minimote/ui/hwhotkeys/HwHotkeysFragment.java
package org.docheinstein.minimote.ui.hwhotkeys;

import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.FragmentActivity;
import androidx.lifecycle.Observer;
import androidx.navigation.fragment.NavHostFragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import org.docheinstein.minimote.R;
import org.docheinstein.minimote.buttons.ButtonType;
import org.docheinstein.minimote.database.DB;
import org.docheinstein.minimote.database.hwhotkey.HwHotkeyEntity;
import org.docheinstein.minimote.ui.base.MinimoteFragment;
import org.docheinstein.minimote.ui.hwhotkey.AddEditHwHotkeyFragment;
import org.docheinstein.minimote.utils.ResUtils;

import java.util.List;

public class HwHotkeysFragment extends MinimoteFragment {

    private static final String TAG = "ServersFragment";

    private HwHotkeyListAdapter uiHwHotkeysListAdapter;
    private RecyclerView.LayoutManager uiHwHotkeyListLayoutManager;
    private RecyclerView uiHwHotkeyList;

    private class HwHotkeyListAdapter extends RecyclerView.Adapter<HwHotkeyListAdapter.HwHotkeyViewHolder> {

        private List<HwHotkeyEntity> mHotkeys;

        public void setHwHotkeys(List<HwHotkeyEntity> hwhotkeys) {
            Log.d(TAG, "Updating hw hotkeys list UI");
            mHotkeys = hwhotkeys;
            notifyDataSetChanged();
        }

        @NonNull
        @Override
        public HwHotkeyViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
            View v = LayoutInflater.from(parent.getContext())
                    .inflate(R.layout.hw_hotkey_item, parent, false);
            return new HwHotkeyViewHolder(v);

        }

        @Override
        public void onBindViewHolder(@NonNull HwHotkeyViewHolder holder, int position) {
            if (mHotkeys == null) {
                Log.w(TAG, "Null hotkeys list");
                return;
            }

            if (position < 0 || position >= mHotkeys.size()) {
                // Fallback
                Log.w(TAG, "Invalid list position (" + position + ")");
                holder.uiHwHotkeyLabel.setText("");
                holder.uiHwHotkeyIcon.setImageIcon(null);
                holder.mRowClickListener = null;
                return;
            }

            final HwHotkeyEntity hwhotkey = mHotkeys.get(position);

            holder.mRowClickListener = new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    Log.v(TAG, "Click on hw hotkey [" + hwhotkey.id + "], opening edit screen");

                    NavHostFragment.findNavController(HwHotkeysFragment.this)
                            .navigate(HwHotkeysFragmentDirections.actionAddEditHwHotkey(hwhotkey.id));
                }
            };
            holder.uiHwHotkeyLabel.setText(getHwHotkeyName(hwhotkey));

            int icon = 0;
            switch (ButtonType.fromString(hwhotkey.button)) {
                case VolumeUp:
                    icon = R.drawable.volume_up;
                    break;
                case VolumeDown:
                    icon = R.drawable.volume_down;
                    break;
                default:
                    Log.w(TAG, "Invalid hw hotkey button (and icon)");
            }

            holder.uiHwHotkeyIcon.setImageResource(icon);
        }


        @Override
        public int getItemCount() {
            return mHotkeys != null ? mHotkeys.size() : 0;
        }


        private String getHwHotkeyName(HwHotkeyEntity hwhotkey) {
            if (hwhotkey == null)
                return "<unknown>";

            StringBuilder sb = new StringBuilder();
            if (hwhotkey.ctrl)
                sb.append(ResUtils.getString(R.string.edit_hotkey_ctrl, getContext()));

            if (hwhotkey.alt) {
                if (sb.length() > 0)
                    sb.append("+");
                sb.append(ResUtils.getString(R.string.edit_hotkey_alt, getContext()));
            }

            if (hwhotkey.altgr) {
                if (sb.length() > 0)
                    sb.append("+");
                sb.append(ResUtils.getString(R.string.edit_hotkey_altgr, getContext()));
            }

            if (hwhotkey.meta) {
                if (sb.length() > 0)
                    sb.append("+");
                sb.append(ResUtils.getString(R.string.edit_hotkey_meta, getContext()));
            }

            if (hwhotkey.shift) {
                if (sb.length() > 0)
                    sb.append("+");
                sb.append(ResUtils.getString(R.string.edit_hotkey_shift, getContext()));
            }

            String key = hwhotkey.key;
            if (sb.length() > 0)
                sb.append("+");
            sb.append(key);

            return sb.toString();
        }

        private class HwHotkeyViewHolder extends RecyclerView.ViewHolder{
            ImageView uiHwHotkeyIcon;
            TextView uiHwHotkeyLabel;
            View.OnClickListener mRowClickListener;

            HwHotkeyViewHolder(@NonNull View itemView) {
                super(itemView);
                uiHwHotkeyIcon = itemView.findViewById(R.id.uiHwHotkeyIcon);
                uiHwHotkeyLabel = itemView.findViewById(R.id.uiHwHotkeyLabel);

                itemView.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        if (mRowClickListener != null)
                            mRowClickListener.onClick(v);
                    }
                });
            }
        }
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setHasOptionsMenu(true);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        final View view = inflater.inflate(R.layout.hw_hotkeys, container, false);

        Log.v(TAG, "HwHotkeysFragment.onCreateView() [" + hashCode() + "]");

        uiHwHotkeyList = view.findViewById(R.id.uiHwHotkeyList);
        uiHwHotkeyList.setHasFixedSize(true);

        uiHwHotkeyListLayoutManager = new LinearLayoutManager(getContext());
        uiHwHotkeyList.setLayoutManager(uiHwHotkeyListLayoutManager);

        uiHwHotkeysListAdapter = new HwHotkeyListAdapter();
        uiHwHotkeyList.setAdapter(uiHwHotkeysListAdapter);

        DB.getInstance().hwhotkeys().getAllObservable().observe(
                getViewLifecycleOwner(), new Observer<List<HwHotkeyEntity>>() {
                    @Override
                    public void onChanged(List<HwHotkeyEntity> hotkeys) {
                        Log.v(TAG, "onChanged() for hw hotkeys");

                        if (hotkeys == null || hotkeys.isEmpty()) {
                            Log.d(TAG, "No hotkeys yet");
                        } else {
                            StringBuilder sb = new StringBuilder("\n");
                            for (HwHotkeyEntity hotkey : hotkeys) {
                                sb.append(">>: ").append(hotkey).append("\n");
                            }
                            Log.i(TAG, sb.toString());
                        }

                        uiHwHotkeysListAdapter.setHwHotkeys(hotkeys);
                    }
                });

        setToolbarTitle("Physical hotkeys");

        return view;
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        Log.d(TAG, "onCreateOptionsMenu");
        inflater.inflate(R.menu.hw_hotkeys, menu);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        switch (item.getItemId()) {
            case R.id.uiAddHwHotkeyMenuItem:
                handleAddHwHotkeyButtonClick();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private void handleAddHwHotkeyButtonClick() {
        FragmentActivity activity = getActivity();

        if (activity == null) {
            Log.w(TAG, "Null activity!?");
            return;
        }

        Log.v(TAG, "Clicked add hw hotkey button");

        NavHostFragment.findNavController(this).navigate(HwHotkeysFragmentDirections
                .actionAddEditHwHotkey(AddEditHwHotkeyFragment.ADD_HW_HOTKEY_MAGIC_ACTION_ID));
    }
}

File: app/src/main/java/org/docheinstein/minimotek/ui/hwhotkeys/HwHotkeysFragment.kt
package org.docheinstein.minimotek.ui.hwhotkeys

import android.app.AlertDialog
import android.os.Bundle
import android.view.*
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.navigation.findNavController
import androidx.navigation.fragment.findNavController
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.snackbar.Snackbar
import dagger.hilt.android.AndroidEntryPoint
import org.docheinstein.minimotek.R
import org.docheinstein.minimotek.buttons.ButtonType
import org.docheinstein.minimotek.database.hotkey.hw.HwHotkey
import org.docheinstein.minimotek.databinding.*
import org.docheinstein.minimotek.ui.base.SelectableListAdapter
import org.docheinstein.minimotek.util.debug
import org.docheinstein.minimotek.util.verbose
import org.docheinstein.minimotek.util.warn

/**
 * Fragment representing the hardware hotkey list
 * (hotkeys mapped to physical buttons, e.g. VolumeUp/VolumeDown).
 * The actions that can be performed on this screen are:
 * - Add a hardware hotkey (opens AddEdit screen)
 * - Edit a hardware hotkey (opens AddEdit screen)
 * - Delete a hardware hotkey
 */

@AndroidEntryPoint
class HwHotkeysFragment : Fragment() {

    private val viewModel: HwHotkeysViewModel by viewModels()
    private lateinit var binding: HwHotkeyListBinding
    private lateinit var adapter: HwHotkeyListAdapter

    private class HwHotkeyDiffCallback : DiffUtil.ItemCallback<HwHotkey>() {
        override fun areItemsTheSame(oldItem: HwHotkey, newItem: HwHotkey): Boolean {
            return oldItem.id == newItem.id
        }

        override fun areContentsTheSame(oldItem: HwHotkey, newItem: HwHotkey): Boolean {
             // UI based equality
            return oldItem.button == newItem.button
        }
    }

    class HwHotkeyListAdapter : SelectableListAdapter<HwHotkey, HwHotkeyListAdapter.ViewHolder>(HwHotkeyDiffCallback()) {
        class ViewHolder(val binding: HwHotkeyListItemBinding) :
                RecyclerView.ViewHolder(binding.root), View.OnCreateContextMenuListener {

            init {
                binding.root.setOnCreateContextMenuListener(this)
            }

            override fun onCreateContextMenu(
                menu: ContextMenu?,
                view: View?,
                context_info: ContextMenu.ContextMenuInfo?
            ) {
                if (view != null && menu != null) {
                    val menuInflater = MenuInflater(view.context)
                    menuInflater.inflate(R.menu.edit_delete, menu)
                    menu.setHeaderTitle(view.context.getString(R.string.choose_an_action))
                }
            }
        }

        override fun doCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
            return ViewHolder(
                HwHotkeyListItemBinding.inflate(
                    LayoutInflater.from(parent.context), parent, false),

            )
        }

        override fun doBindViewHolder(holder: ViewHolder, position: Int) {
            val hwHotkey = getItem(position)

            // Text
            holder.binding.label.text = hwHotkey.button.name

            // Icon
            val icon = when (hwHotkey.button) {
                ButtonType.VolumeDown -> { R.drawable.volume_down }
                ButtonType.VolumeUp -> { R.drawable.volume_up }
            }

            holder.binding.icon.setImageResource(icon)

            // Click listener: open AddEditHwhotkey
            holder.binding.root.setOnClickListener {
                debug("Click on hwHotkey $hwHotkey")
                 holder.itemView.findNavController().navigate(
                    HwHotkeysFragmentDirections.actionAddEditHwHotkey(
                        hwHotkey.id,
                        holder.itemView.context.getString(R.string.toolbar_title_edit_hw_hotkey)
                ))
            }
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        verbose("HwHotkeysFragment.onCreate()")
        super.onCreate(savedInstanceState)
        setHasOptionsMenu(true)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?,
                              savedInstanceState: Bundle?): View? {
        verbose("HwHotkeysFragment.onCreateView()")

        binding = HwHotkeyListBinding.inflate(inflater, container, false)

        // Hardware hotkeys adapter
        adapter = HwHotkeyListAdapter()
        binding.hwHotkeyList.adapter = adapter

        // Observe hardware hotkeys changes
        viewModel.hwHotkeys.observe(viewLifecycleOwner) { hwHotkeys ->
            debug("Hardware hotkeys update received in UI (size = ${hwHotkeys.size})")
            adapter.submitList(hwHotkeys)
        }

        return binding.root
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.add, menu)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.add_menu_item -> {
                handleAddAction()
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onContextItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.edit_menu_item -> {
                handleEditAction()
                return true
            }
            R.id.delete_menu_item -> {
                handleDeleteAction()
                return true
            }
        }
        return super.onContextItemSelected(item)
    }

    private fun handleAddAction() {
        verbose("HwHotkeysFragment.handleAddAction()")

        findNavController().navigate(
            HwHotkeysFragmentDirections.actionAddEditHwHotkey(
                AddEditHwHotkeyViewModel.HW_HOTKEY_ID_NONE,
                getString(R.string.toolbar_title_add_hw_hotkey)
            )
        )
    }

    private fun handleEditAction() {
        verbose("HwHotkeysFragment.handleEditAction()")

        val hwHotkey = adapter.selectedItem
        debug("Going to edit hardware hotkey at position ${adapter.selectedPosition}: ${hwHotkey?.id}")
        if (hwHotkey != null) {
            findNavController().navigate(
                HwHotkeysFragmentDirections.actionAddEditHwHotkey(
                    hwHotkey.id,
                    getString(R.string.toolbar_title_edit_hw_hotkey)
                )
            )
        } else {
            warn("No hotkey selected, cannot edit")
        }
    }

    private fun handleDeleteAction() {
        verbose("HwHotkeysFragment.handleDeleteAction()")

        val hwHotkey = adapter.selectedItem
        debug("Going to delete hardware hotkey at position ${adapter.selectedPosition}: ${hwHotkey?.id}")
        if (hwHotkey != null) {
            AlertDialog.Builder(requireActivity())
                .setTitle(R.string.delete_hw_hotkey_confirmation_title)
                .setMessage(getString(R.string.delete_hw_hotkey_confirmation_message, hwHotkey.button.name))
                .setPositiveButton(R.string.ok) { _, _ ->
                    // Actually delete
                    viewModel.delete(hwHotkey.id)
                    Snackbar.make(
                        requireParentFragment().requireView(),
                        getString(R.string.removed, hwHotkey.button.name),
                        Snackbar.LENGTH_LONG
                    ).show()
                }
                .setNegativeButton(R.string.cancel, null)
                .show()
        }
    }
}